### PR TITLE
Implement package signing with swift-certificates

### PR DESCRIPTION
**Pull request by yim-lee**
This PR focuses on signing (i.e., generating signature). It can validate signature but it doesn't yet honor user configuration. That will be done in a follow-up PR.

### CODE REVIEW

File: Sources/PackageRegistryTool/PackageRegistryTool+Sign.swift [hunk 1]
```diff
@@ -18,6 +18,7 @@ import Basics
 import CoreCommands
 import PackageSigning
 import TSCBasic
+@_implementationOnly import X509 // FIXME: need this import or else SwiftSigningIdentity init at L128 fails
```
Comment (yim-lee) [inline Sources/PackageRegistryTool/PackageRegistryTool+Sign.swift:21]: 👀

**Comment by yim-lee** [comment]
cc @Lukasa @dnadoba

**Comment by yim-lee** [comment]
This is ready for review but can't be merged until we sort out CI and https://github.com/apple/swift-certificates/pull/23 is merged.

File: Sources/PackageSigning/SignatureProvider.swift [hunk 2]
```diff
@@ -193,86 +201,61 @@ struct CMSSignatureProvider: SignatureProviderProtocol {
...
+            if failure.validationFailures.isEmpty {
+                return .certificateNotTrusted
+            } else {
+                return .certificateInvalid("\(failure.validationFailures)") // TODO: format error message
+            }
+        case .invalidCMSBlock(let error):
+            return .invalid(error.reason)
         }
+    }
+}
```
Comment (tomerd) [inline Sources/PackageSigning/SignatureProvider.swift:229]: very nice cleanup

File: Sources/PackageSigning/SigningEntity/SigningEntity.swift [hunk 3]
```diff
@@ -79,21 +55,18 @@ public struct SigningEntity: Hashable, Codable, CustomStringConvertible {
...
-            [SigningEntityType.oid_adpSwiftPackageMarker as CFString] as CFArray,
-            nil
-        ) as? [CFString: Any] else {
-            return nil
+        // TODO: check that cert is chained to WWDR roots
+        if self.hasExtension(oid: ASN1ObjectIdentifier.NameAttributes.adpSwiftPackageMarker) {
+            return .adp
         }
-        return dict.isEmpty ? nil : .adp
+        return nil
```
Comment (tomerd) [inline Sources/PackageSigning/SigningEntity/SigningEntity.swift:70]: nice cleanup

File: Sources/PackageSigning/SigningIdentity.swift [hunk 4]
```diff
@@ -12,100 +12,53 @@
 
 import struct Foundation.Data
 
-#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+#if canImport(Darwin)
```
Comment (tomerd) [inline Sources/PackageSigning/SigningIdentity.swift:15]: 🙇

File: Tests/CommandsTests/PackageRegistryToolTests.swift [hunk 5]
```diff
@@ -421,26 +458,36 @@ final class PackageRegistryToolTests: CommandsTestCase {
...
         try await withTemporaryDirectory { temporaryDirectory in
             let archivePath = temporaryDirectory.appending(component: "fake.zip")
             try localFileSystem.writeFileContents(archivePath, string: "test")
 
-            let privateKeyPath = temporaryDirectory.appending(component: "private-key")
-            // TODO: write the private key
-            let certificatePath = temporaryDirectory.appending(component: "certificate")
-            // TODO: write the cert
+            let certificatePath = temporaryDirectory.appending(component: "certificate.cer")
+            let privateKeyPath = temporaryDirectory.appending(component: "private-key.p8")
```
Comment (tomerd) [inline Tests/CommandsTests/PackageRegistryToolTests.swift:475]: <3

File: Tests/CommandsTests/PackageRegistryToolTests.swift [hunk 6]
```diff
@@ -449,10 +496,24 @@ final class PackageRegistryToolTests: CommandsTestCase {
...
+                of: signature,
+                for: archive,
+                in: .cms_1_0_0,
+                verifierConfiguration: verifierConfiguration,
+                observabilityScope: observabilityScope
+            )
+            guard case .valid = signatureStatus else {
+                return XCTFail("Expected signature status to be .valid but got \(signatureStatus)")
+            }
+        }
```
Comment (tomerd) [inline Tests/CommandsTests/PackageRegistryToolTests.swift:516]: <3

File: Sources/PackageRegistryTool/PackageRegistryTool+Sign.swift (hunk 1 above)
Comment (yim-lee) [reply Sources/PackageRegistryTool/PackageRegistryTool+Sign.swift:21]: I thought it was caused by the code leaking X509 error types but addressing that doesn't fix this

File: Sources/PackageRegistry/RegistryClient.swift [hunk 7]
```diff
@@ -1061,6 +1062,20 @@ public final class RegistryClient: Cancellable {
         """.utf8)
         body.append(packageArchiveContent)
 
+        if let signature = signature {
```
Comment (yim-lee) [inline Sources/PackageRegistry/RegistryClient.swift:1065]: Start sending signature

**Review by Lukasa — APPROVED** [review]
Looks awesome!

File: Sources/PackageSigning/SignatureProvider.swift [hunk 8]
```diff
@@ -78,201 +82,189 @@ public enum SignatureStatus: Equatable {
...
+        content: [UInt8],
+        verifierConfiguration: VerifierConfiguration,
+        observabilityScope: ObservabilityScope
+    ) async throws -> SignatureStatus {
+        let result = await CMS.isValidSignature(
+            dataBytes: content,
+            signatureBytes: signature,
+            trustRoots: CertificateStore(
+                try verifierConfiguration.trustedRoots
+                    .map { try Certificate(derEncoded: $0) }
```
Comment (dnadoba) [inline Sources/PackageSigning/SignatureProvider.swift:218]: parsing the trusted roots can be an expensive operation. Should we do some caching if we need to validate multiple signatures with the same verifier configuration? cc: @Lukasa

**Review by dnadoba — APPROVED** [review]
LGTM

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

**Comment by yim-lee** [comment]
@swift-ci test Windows platform

**Comment by yim-lee** [comment]
@swift-ci test Windows platform

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

File: Sources/PackageSigning/SignatureProvider.swift (hunk 8 above)
Comment (yim-lee) [reply Sources/PackageSigning/SignatureProvider.swift]: added caching

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

**Review by neonichu — APPROVED** [review]

**Review by tomerd — APPROVED** [review]
very nice

**Comment by neonichu** [comment]
We need https://github.com/apple/swift/pull/64157 to fix the Windows failure.

**Pull request merged**